cyipopt/cython/ipopt_wrapper.pyx

```diff
@@ -654,6 +654,102 @@ cdef class Problem:
 
         return np_x, info
 
+    def get_current_iterate(self, scaled=False):
+        # Allocate arrays to hold the current iterate
+        cdef np.ndarray[DTYPEd_t, ndim=1] np_x
+        cdef np.ndarray[DTYPEd_t, ndim=1] np_mult_x_L
+        cdef np.ndarray[DTYPEd_t, ndim=1] np_mult_x_U
+        cdef np.ndarray[DTYPEd_t, ndim=1] np_g
+        cdef np.ndarray[DTYPEd_t, ndim=1] np_mult_g
+        np_x = np.zeros((self.__n,), dtype=DTYPEd).flatten()
+        np_mult_x_L = np.zeros((self.__n,), dtype=DTYPEd).flatten()
+        np_mult_x_U = np.zeros((self.__n,), dtype=DTYPEd).flatten()
+        np_g = np.zeros((self.__m,), dtype=DTYPEd).flatten()
+        np_mult_g = np.zeros((self.__m,), dtype=DTYPEd).flatten()
+
+        # Cast to C data types
+        x = <Number*>np_x.data
+        mult_x_L = <Number*>np_mult_x_L.data
+        mult_x_U = <Number*>np_mult_x_U.data
+        g = <Number*>np_g.data
+        mult_g = <Number*>np_mult_g.data
+
+        # NOTE: GetIpoptCurrentIterate can *only* be called during an
+        # intermediate callback (otherwise __nlp->tnlp->ip_data_ is NULL)
+        # TODO: Either catch error or avoid calling if we are not in an
+        # intermediate callback
```

Comment:
Since the ipopt function is only intended to run during the intermediate callback, then I think checking if `ip_data_` is NULL and raising an error here is a good choice.

Reply:
I agree that this would be the best way to do this, but I'm not 100% sure how to access `ip_data_` from Cython, as `tnlp` is an instance of the c++ class `StdInterfaceTNLP`. I'll give this a shot, but unless somebody with more cython knowledge than me knows how, we have to use an alternative.

Reply:
I wonder how ipopt handles this in C. Shouldn't it have the same issue? If you call one of the functions outside of the intermediate callback?

Reply:
I looked into this a little more, and the problem seems to be that `__nlp->tnlp` is either undefined or NULL if we call before or after `IpoptSolve`. This happens when calling from C as well. I've submitted a patch to fix this in Ipopt by checking `tnlp` before accessing it (https://github.com/coin-or/Ipopt/pull/645). 

When this gets fixed in Ipopt, we can just fall back on whatever we decide to do when the return value is 0. Until then, I think the easiest stopgap is to have a global flag indicating whether we are in a callback. It would be nicer to check `IsNull(__nlp->tnlp)`, but this is a c++ function called on a SmartPtr-to-c++ class, which I don't know how to do in Cython.

Reply:
Sounds good. I'm fine with a flag.